wallet/wallet.go

```diff
@@ -0,0 +1,522 @@
+	}
+
+	poolTxns := sw.cm.PoolTransactions()
+
+	var annotated []Transaction
+	for _, txn := range poolTxns {
+		wt := Transaction{
+			ID:          txn.ID(),
+			Transaction: txn,
+			Source:      TxnSourceTransaction,
+			Timestamp:   time.Now(),
+		}
+
+		for _, sci := range txn.SiacoinInputs {
+			if sco, ok := utxos[types.Hash256(sci.ParentID)]; ok {
+				wt.Outflow = wt.Outflow.Add(sco.Value)
+			}
+		}
+
+		for i, sco := range txn.SiacoinOutputs {
+			if sco.Address == sw.addr {
+				wt.Inflow = wt.Inflow.Add(sco.Value)
+				utxos[types.Hash256(txn.SiacoinOutputID(i))] = sco
+			}
+		}
+
+		if wt.Inflow.IsZero() && wt.Outflow.IsZero() {
+			continue
+		}
+
+		annotated = append(annotated, wt)
+	}
+	return annotated, nil
+}
+
+// IsRelevantTransaction returns true if the v1 transaction is relevant to the
+// address
+func IsRelevantTransaction(txn types.Transaction, addr types.Address) bool {
+	for _, sci := range txn.SiacoinInputs {
+		if sci.UnlockConditions.UnlockHash() == addr {
+			return true
+		}
+	}
+
+	for _, sco := range txn.SiacoinOutputs {
+		if sco.Address == addr {
+			return true
+		}
+	}
+
+	for _, sci := range txn.SiafundInputs {
+		if sci.UnlockConditions.UnlockHash() == addr {
+			return true
+		}
+	}
+
+	for _, sfo := range txn.SiafundOutputs {
+		if sfo.Address == addr {
+			return true
+		}
+	}
+	return false
+}
+
+// NewSingleAddressWallet returns a new SingleAddressWallet using the provided private key and store.
+func NewSingleAddressWallet(priv types.PrivateKey, cm ChainManager, store SingleAddressStore, log *zap.Logger) (*SingleAddressWallet, error) {
+	tip, err := store.LastIndexedTip()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get last wallet change: %w", err)
+	}
+
+	seedHash := types.HashBytes(priv[:])
+	if err := store.VerifyWalletKey(seedHash); errors.Is(err, ErrDifferentSeed) {
+		tip = types.ChainIndex{}
+		if err := store.ResetWallet(seedHash); err != nil {
+			return nil, fmt.Errorf("failed to reset wallet: %w", err)
+		}
+		log.Info("wallet reset due to seed change")
+	} else if err != nil {
+		return nil, fmt.Errorf("failed to verify wallet key: %w", err)
+	}
+
+	sw := &SingleAddressWallet{
+		priv: priv,
+		tip:  tip,
+
+		store: store,
+		cm:    cm,
+		log:   log,
+
+		addr:   types.StandardUnlockHash(priv.PublicKey()),
+		locked: make(map[types.Hash256]bool),
+	}
+
+	go func() {
+		// note: start in goroutine to avoid blocking startup
+		err := cm.AddSubscriber(store, tip)
+		if err != nil { // no way to check for specific reorg error, so reset everything
+			sw.log.Warn("rescanning blockchain due to unknown consensus change ID")
+			// reset change ID and subscribe again
+			if err := store.ResetWallet(seedHash); err != nil {
+				sw.log.Fatal("failed to reset wallet", zap.Error(err))
+			} else if err = cm.AddSubscriber(store, types.ChainIndex{}); err != nil {
+				sw.log.Fatal("failed to reset consensus change subscription", zap.Error(err))
+			}
+		}
+	}()
```

Comment:
Can we remove subscribing the `store` from `NewSingleAddressWallet` and expect the user to pass in an already subscribed store which resets itself in case there is some inconsistency?

Because our store in the renter subscribes for more than just the wallet so it would be a bit awkward to have the following

```go
store := newSQLStore() // serves as object store, wallet store, host store etc.

saw := newSingleAddressWallet(store) // subscribes store to consensus

b := newBus(saw, store)
```

That would also get rid of `ResetWallet` from the interface if we just expect the caller to handle an invalid subscription when creating the store. Then it would look more like

```go
store := newSQLStore()

subscribeToConsensus(store)
if err {
  store.ResetSubscription()
}

saw := newSingleAddressWallet(store)
```